### PR TITLE
Make swiftToolchainVersion optional for ObjC only frameworks

### PR DESCRIPTION
**Pull request by kenji21**
Fix issue #2677

### CODE REVIEW

**Comment by blender** [comment]
Code looks good to me. Can you show me how this is change affects the decision to rebuild (or not) the dependencies?

**Comment by kenji21** [comment]
replace spaces by tabs (IDEs should auto detect indentation characters used...)

**Comment by kenji21** [comment]
The case "not a swift framework" is already managed : https://github.com/Carthage/Carthage/blob/master/Source/CarthageKit/VersionFile.swift#L149

If an objc only framework is updated to use swift (or converted entirely to), I assume a new (major) version will be created

I've added a test : 
https://github.com/Carthage/Carthage/pull/2707/files#diff-75f7a281ef0b6a7f62a9b105850dda7bR168

**Comment by hlineholm** [comment]
I've created an alternative (#2708) which does make use of [Xcode.swift:isSwiftFramework](https://github.com/Carthage/Carthage/blob/0aee76db5e5172edb1c618e4d303acfd1af741c2/Source/CarthageKit/Xcode.swift#L119)

**Comment by blender** [comment]
I think #2708 is a little cleaner. I would merge this one first and then @hlineholm 's patch.

So in the end we're not adding a property to the .version file? We're just relying on a missing property?

**Comment by kenji21** [comment]
> So in the end we're not adding a property to the .version file? We're just relying on a missing property?

Yes, when I tried to add a boolean, Codable needs a value when creating VersionFile from a file, here is a "playgroundable" code:

```
struct CachedFramework: Codable {
	enum CodingKeys: String, CodingKey {
		case name = "name"
		case hash = "hash"
		case swiftToolchainVersion = "swiftToolchainVersion"
		case isSwiftFramework = "isSwiftFramework"
	}
	
	let name: String
	let hash: String
	let swiftToolchainVersion: String?
	var isSwiftFramework: Bool
}

let json = "{\"name\":\"ObjcFramework\",\"hash\":\"Hash\",\"swiftToolchainVersion\"}"
let jsonData = json.data(using: .utf8)!
let cachedFramework = try? JSONDecoder().decode(CachedFramework.self, from: jsonData)

if cachedFramework == nil {
	print("not backward compatible")
}
```

which prints `not backward compatible` and I don't know how to nicely fix this (specifying a default value for the isSwiftFramework to true didn't help)

**Comment by kenji21** [comment]
Tried with the Cartfile/commands of issue #2677:

```
$ carthage version
0.32.0
$ carthage update --platform iOS --no-use-binaries --use-submodules
*** Fetching SwiftiumKit
*** Fetching Archimedes
*** Checking out Archimedes at "1.1.5"
*** Checking out SwiftiumKit at "v1.4.0"
*** xcodebuild output can be found in /var/folders/7v/dfg0p2vj1z58n65d99l2n1lh0000gp/T/carthage-xcodebuild.JJ2yn1.log
*** Building scheme "Archimedes iOS" in Archimedes.xcworkspace
*** Building scheme "SwiftiumKit" in SwiftiumKit.xcodeproj
$ ls -a Carthage/Build/
.                    ..                   .Archimedes.version  .SwiftiumKit.version iOS
$ cat Carthage/Build/.Archimedes.version
{
  "commitish" : "1.1.5",
  "iOS" : [
    {
      "name" : "Archimedes",
      "hash" : "d2cdb749ef6bcf813c7973ce8ba93e3edd2ac2e18da6f5315c4bf76fedec6e98"
    }
  ]
}
$ cat Carthage/Build/.SwiftiumKit.version
{
  "commitish" : "v1.4.0",
  "iOS" : [
    {
      "name" : "SwiftiumKit",
      "hash" : "4f7a7bb95c913c7549e2847fa6f335898ea24aabd4c10a4f52ad1b58ad704fb1"
    }
  ]
}
```

Swift framework doesn't have the `swiftToolchainVersion` json key
Now let's bootstrap on top of 0.32.0 cache:
```
~/github/Carthage/.build/x86_64-apple-macosx10.10/release/carthage bootstrap --platform iOS --no-use-binaries --use-submodules --cache-builds
*** Checking out SwiftiumKit at "v1.4.0"
*** Checking out Archimedes at "1.1.5"
*** xcodebuild output can be found in /var/folders/7v/dfg0p2vj1z58n65d99l2n1lh0000gp/T/carthage-xcodebuild.y48rty.log
*** Valid cache found for Archimedes, skipping build
*** Valid cache found for SwiftiumKit, skipping build
```

The cache is correct so no dependencies recompilation needed.

File: Source/CarthageKit/Xcode.swift [hunk 1]
```diff
@@ -50,6 +50,16 @@ private func parseSwiftVersionCommand(output: String?) -> String? {
...
 }
 
+/// Determines the Swift version of a framework at a given `URL`.
+internal func frameworkSwiftVersionIfIsSwiftFramework(_ frameworkURL: URL) -> SignalProducer<String?, SwiftVersionError> {
+	guard isSwiftFramework(frameworkURL) else {
+		return SignalProducer(value: nil)
+	}
+	return frameworkSwiftVersion(frameworkURL).flatMap(.merge, { (swiftVersion) -> SignalProducer<String?, SwiftVersionError> in
+		return SignalProducer(value: swiftVersion)
+	})
```
Comment (mdiep) [inline Source/CarthageKit/Xcode.swift]: This can be `frameworkSwiftVersion(frameworkURL).map(Optional.some)`

**Comment by mdiep** [comment]
This looks good to me. I'd just like to see 1 minor change: using `map` instead of `flatMap`. I think that's a lot clearer.

File: Source/CarthageKit/Xcode.swift (hunk 1 above)
Comment (kenji21) [reply Source/CarthageKit/Xcode.swift]: Ammended my commit to simplify code : https://github.com/Carthage/Carthage/pull/2707/commits/602e99905e95284fde8d1f472bf6fbd622c1cfcf#diff-f66644e8cfbcad3c3826395a690ddfbcR58

Thanks

**Comment by mdiep** [comment]
Thank you for the PR!

**Pull request merged**